naga/src/proc/constant_evaluator.rs

```diff
@@ -2196,3 +2275,46 @@ impl TryFromAbstract<f64> for u32 {
         Err(ConstantEvaluatorError::AutomaticConversionFloatToInt { to_type: "u32" })
     }
 }
+
+impl From<f64> for Expression {
+    fn from(value: f64) -> Self {
+        Self::Literal(Literal::AbstractFloat(value))
+    }
+}
+
+impl From<f32> for Expression {
+    fn from(value: f32) -> Self {
+        Self::Literal(Literal::F32(value))
+    }
+}
+
+impl From<i64> for Expression {
+    fn from(value: i64) -> Self {
+        Self::Literal(Literal::AbstractInt(value))
+    }
+}
+
+impl From<u32> for Expression {
+    fn from(value: u32) -> Self {
+        Self::Literal(Literal::U32(value))
+    }
+}
+
+impl From<i32> for Expression {
+    fn from(value: i32) -> Self {
+        Self::Literal(Literal::I32(value))
+    }
+}
```

Comment:
- [ ] issue: I am not inclined to merely assume that these `From` implementations are okay to add; I suspect that they would have been added a long time ago, if they were. Some of the brevity of the built-ins I'm contributing here depends on these currently. Wanted to ask @jimblandy's thoughts about alternatives here.

    Personally, I actually really like these, minus maybe `From<{u,f}64>`. I don't think it's obvious that these would correspond to `Abstract{Int,Float}`.

Reply:
> Personally, I actually really like these, minus maybe `From<{u,f}64>`. I don't think it's obvious that these would correspond to `Abstract{Int,Float}`.

I think having these cleans up the code nicely but I think that we shouldn't have them for `{i,f}64` -> `Abstract{Int,Float}` though. Maybe we should introduce 2 wrapper types for `Abstract{Int,Float}` around `{i,f}64`.

Reply:
So, I just did a bunch of stuff with the `enum`s subsetting `Literal` that have been added by this PR:

* I now use the `gen_component_wise_extractor` macro to generate these, with largely the same input as before. The only thing I had to add to the current macro input is the concrete types that will be stored in the subset `enum`.
* I have removed the `From` implementations on primitive types behind the controversy in this discussion. Instead, I now use `impl From<$target<1>> for Expression` inside the macro (i.e., `Scalar<1>`). I chose to compromise a bit on syntax to do this: you now have to return a `[T; 1]` instead of just a `T`.

    I suspect we may actually want sizes greater than 1 later, but I'm not sure. If not, this shouldn't be hard to change back to requiring "just" `T` later.
* Not super relevant for this convo, but FTR: I have also removed `pub(super)` visibility. It wasn't necessary.

I think this pretty neatly resolves this conversation. The `abs` and `acos` built-ins showcase how code looks now:

```rust
            crate::MathFunction::Abs => {
                component_wise_scalar(self, span, [arg], |args| match args {
                    Scalar::AbstractFloat([e]) => Ok(Scalar::AbstractFloat([e.abs()])),
                    Scalar::F32([e]) => Ok(Scalar::F32([e.abs()])),
                    Scalar::AbstractInt([e]) => Ok(Scalar::AbstractInt([e.abs()])),
                    Scalar::I32([e]) => Ok(Scalar::I32([e.wrapping_abs()])),
                    Scalar::U32([e]) => Ok(Scalar::U32([e])), // TODO: just re-use the expression, ezpz
                })
            }
            crate::MathFunction::Acos => {
                component_wise_float!(self, span, [arg], |e| { Ok([e.acos()]) })
            }
```

Still pretty readable, methinks!

Reply:
> I suspect we may actually want sizes greater than 1 later, but I'm not sure. If not, this shouldn't be hard to change back to requiring "just" `T` later.

I don't think we will, or I hope we don't as scalars should only hold one value.
It's a bit odd to see the values being wrapped in an array with length 1.

@jimblandy thoughts on this?

Reply:
@teoxoy: I think we might need to improve the naming of these convenience structures to handle your concern? `Float` and `Scalar`  are essentially a representation of components. They could be a single scalar, or multiple components of a `vec`. Having more than 1 would "just" mean that we want to emi a `vec` and not just a single scalar value.

Reply:
@teoxoy, @jimblandy: Ping on this.

Reply:
Ok, I think this is fine for now, we can revisit it later.